bls/src/main/java/tech/pegasys/teku/bls/hashToG2/Helper.java

```diff
@@ -62,65 +66,94 @@ static boolean isOnCurve(JacobianPoint p) {
    * @return true if the point is in G2, false otherwise
    */
   static boolean isInG2(JacobianPoint p) {
-    return isOnCurve(p) && g2_map_to_infinity(p).isInfinity();
+    return isOnCurve(p) && mapG2ToInfinity(p).isInfinity();
   }
 
   /**
-   * Apply a transformation that maps G2 elements, and only G2 elements, to infinity
+   * Apply a transformation that maps G2 elements, and only G2 elements, to infinity.
    *
    * <p>Uses the technique from https://eprint.iacr.org/2019/814.pdf section 3.1
    *
-   * @return
+   * @param p the point on the curve to test
+   * @return the point at infinity iff p is in G2, otherwise an arbitrary point
    */
-  static JacobianPoint g2_map_to_infinity(JacobianPoint p) {
+  @VisibleForTesting
+  static JacobianPoint mapG2ToInfinity(JacobianPoint p) {
     JacobianPoint psi1 = psi(p);
     JacobianPoint psi2 = psi(psi1);
     JacobianPoint psi3 = psi(psi2);
     return mxChain(psi3).add(psi2).neg().add(p);
   }
 
   /**
-   * Hashes a string msg of any length into an element of the FP2 field.
+   * Produces a uniformly random byte string of arbitrary length using SHA-256.
+   *
+   * <p>As defined at https://tools.ietf.org/html/draft-irtf-cfrg-hash-to-curve-07#section-5.3.1
    *
-   * <p>As defined at https://tools.ietf.org/html/draft-irtf-cfrg-hash-to-curve-04#section-5.3
+   * @param message the message to hash
+   * @param dst the domain separation tag for the cipher suite
+   * @param lengthInBytes the number of bytes we want to obtain
+   * @return a uniformly random sequence of Bytes
+   */
+  @VisibleForTesting
+  static Bytes expandMessage(Bytes message, Bytes dst, int lengthInBytes) {
+    checkArgument(dst.size() < 256, "The DST must be 255 bytes or fewer.");
+    checkArgument(lengthInBytes > 0, "Number of bytes requested must be greater than zero.");
+
+    final int ell = 1 + (lengthInBytes - 1) / SHA256_HASH_SIZE;
+    checkArgument(ell <= 255, "Too many bytes of output were requested.");
+
+    byte[] uniformBytes = new byte[ell * SHA256_HASH_SIZE];
+
+    Bytes dstPrime = Bytes.concatenate(dst, Bytes.of((byte) dst.size()));
+    Bytes zPad = Bytes.wrap(new byte[SHA256_BLOCK_SIZE]);
+    Bytes libStr = Bytes.ofUnsignedShort(lengthInBytes);
+    Bytes b0 =
+        Hash.sha2_256(Bytes.concatenate(zPad, message, libStr, Bytes.of((byte) 0), dstPrime));
+    Bytes bb = Hash.sha2_256(Bytes.concatenate(b0, Bytes.of((byte) 1), dstPrime));
+    System.arraycopy(bb.toArrayUnsafe(), 0, uniformBytes, 0, SHA256_HASH_SIZE);
+    for (int i = 1; i < ell; i++) {
+      bb = Hash.sha2_256(Bytes.concatenate(b0.xor(bb), Bytes.of((byte) (i + 1)), dstPrime));
```

Comment:
The spec seems to suggest that you should store the intermediate `b_i` values, concatenate them together, and return the first `len_in_bytes`. But we seem to be only returning the first `len_in_bytes` of `b_ell`. Am I missing something?

Reply:
The next line concatenates `bb` onto the `uniformBytes` array, so it's equivalent.

This is an optimisation - the way the Bytes library does concatenation we'd end up re-copying the whole of `uniformBytes` every time we added the next chunk, giving us O(n^2/2) copies to do rather than O(n) as we have here, where n is the number of chunks. It's not very important in the grand scheme of things as n is not large. But I find it more pleasing 😄